renku_notebooks/api/classes/server.py

```diff
@@ -161,219 +166,439 @@ def _get_image(self, image):
+        stateful_set_container_modifications = []
+        # Add labels and annotations - applied to overall manifest and secret only
+        labels = {
+            "app": "jupyterhub",
+            "component": "singleuser-server",
+            f"{current_app.config['RENKU_ANNOTATION_PREFIX']}commit-sha": self.commit_sha,
+            f"{current_app.config['RENKU_ANNOTATION_PREFIX']}gitlabProjectId": str(
+                self.gl_project.id
+            ),
+            current_app.config["RENKU_ANNOTATION_PREFIX"]
+            + "safe-username": self._user.safe_username,
+        }
+        annotations = {
+            f"{current_app.config['RENKU_ANNOTATION_PREFIX']}commit-sha": self.commit_sha,
+            f"{current_app.config['RENKU_ANNOTATION_PREFIX']}gitlabProjectId": str(
+                self.gl_project.id
+            ),
+            current_app.config["RENKU_ANNOTATION_PREFIX"]
+            + "safe-username": self._user.safe_username,
+            f"{current_app.config['RENKU_ANNOTATION_PREFIX']}username": self._user.username,
+            f"{current_app.config['RENKU_ANNOTATION_PREFIX']}servername": self.server_name,
+            f"{current_app.config['RENKU_ANNOTATION_PREFIX']}branch": self.branch,
+            f"{current_app.config['RENKU_ANNOTATION_PREFIX']}git-host": self.git_host,
+            f"{current_app.config['RENKU_ANNOTATION_PREFIX']}namespace": self.gl_project.namespace[
+                "full_path"
+            ],
+            current_app.config["RENKU_ANNOTATION_PREFIX"]
+            + "projectName": self.gl_project.name.lower(),
+            f"{current_app.config['RENKU_ANNOTATION_PREFIX']}requested-image": self.image,
+        }
+        # Add image pull secret if image is private
+        if is_image_private:
+            image_pull_secret_name = self.server_name + "-secret"
+            extra_resources.append(
+                {
+                    "api": "CoreV1Api",
+                    "creationMethod": "create_namespaced_secret",
+                    "resourceSpec": {
+                        "apiVersion": "v1",
+                        "data": {".dockerconfigjson": self._get_registry_secret()},
+                        "kind": "Secret",
+                        "metadata": {
+                            "name": image_pull_secret_name,
+                            "namespace": self._k8s_namespace,
+                            "labels": labels,
+                            "annotations": annotations,
+                        },
+                        "type": "kubernetes.io/dockerconfigjson",
+                    },
+                }
+            )
+            stateful_set_image_pull_secret_modifications.append(
+                {"name": image_pull_secret_name}
+            )
+        # Add git init / sidecar container
+        stateful_set_container_modifications.append(
+            {
+                "image": current_app.config["GIT_SIDECAR_IMAGE"],
```

Comment:
Ok so this is def a problem. I added a sleep command in the bash script that clones the repository and then starts the git-sidecar. And I can definitely get into a session that has not cloned its repository. Even though the sleep command waits for the git server and and git proxy to become available it sleeps before starting to clone the repo. And the session does indeed become available to the user even though the repo has not been cloned. I will try to setup the readiness and liveness probes in k8s and see if this helps. Otherwise we will go with the solution that @rokroskar proposed here.